haskell-style.md

```diff
@@ -660,3 +660,329 @@ module Driver
     ) where
 
   ```
+
+## Haddocks
+
+Haskell modules should be commented with valid Haddock documentation. We are not
+yet requiring a certain level of coverage, but it is strongly encouraged.
+
+### General guides
+
+- Use proper Haddock [markup][]
+- Link all identifiers, anywhere they appear
+
+  Bad
+
+  ```hs
+  -- | Construct a @'FlipFlop'@
+  --
+  -- If the size is right, you will get a @Right FlipFlop@, otherwise a @Left@
+  --
+  ```
+
+  Good
+
+  ```hs
+  -- | Construct a @'FlipFlop'@
+  --
+  -- If the size is right, you will get a @'Right' 'FlipFlop'@, otherwise a
+  -- @'Left'@.
+  --
+  ```
+
+- Use leading documentation (`-- |`) for top-level definitions and trailing
+  documentation (`-- ^`) for record attributes and function arguments
+
+  Bad
+
+  ```hs
+  data Foo = Foo
+    {
+    -- | Foo's foo
+      fooFoo :: Foo
+    -- | Foo's bar
+    , fooBar :: Bar
+    }
+  -- ^ What even is a Foo?
+  ```
+
+  Good
+
+  ```hs
+  -- | What even is a Foo?
+  data Foo = Foo
+    { fooFoo :: Foo
+    -- ^ Foo's foo
```

Comment:
Do we want to indent these comments by 2, so they aren't at a lower indentation than the thing they are describing?

```
data Foo
  = Foo
  { fooFoo :: Foo
    -- ^ Foo's foo
  ...
```

I'm fine either way on this, just throwing it out there.

Reply:
I've never seen that done that way, so I'm generally against sprouting our own convention here that differs with what you see in the wild. Then again, I haven't done an exhaustive search (or any really) so maybe that is a common way to do it?

Reply:
Never? I thought this was by far the most common

* `lens`: http://hackage.haskell.org/package/lens-4.16.1/docs/src/Control-Lens-TH.html (gotta ctrl + F for `-- ^`
* `aeson`: http://hackage.haskell.org/package/aeson-1.4.0.0/docs/src/Data.Aeson.Types.Internal.html#Options
* `shake` actually goes further than aligned https://hackage.haskell.org/package/shake-0.16.4/docs/src/Development.Shake.Internal.Options.html#ShakeOptions
* `time` is using the `-- |` style, but still aligning https://hackage.haskell.org/package/time-1.9.1/docs/src/Data-Time-LocalTime-Internal-TimeZone.html#TimeZone

I wouldn't call it "sprouting our own convention" at all.

Reply:
- Lens

  This is doing exactly what I'm proposing, aligning with `,` with a weird first element.

   ```hs
   module Control.Lens.TH
    (
    -- * Constructing Lenses Automatically
    -- ** Lenses for data fields
      makeLenses, makeLensesFor
    , makeClassy, makeClassyFor, makeClassy_
    , makeFields
    , makeFieldsNoPrefix
    -- ** Prisms
    , makePrisms
    , makeClassyPrisms
    -- ** Wrapped
    , makeWrapped
  ```

- Aeson is doing what you suggest, :ok_hand: 

- Shake is doing what you suggest, if you take a liberal view since it doesn't put a space after `,` and is not aligning to identifier

  ```hs
    {shakeFiles :: FilePath
          -- ^ Defaults to @.shake@. The directory used for storing Shake metadata files.
          --   All metadata files will be named @'shakeFiles'\/.shake./file-name/@, for some @/file-name/@.
          --   If the 'shakeFiles' directory does not exist it will be created.
    ,shakeThreads :: Int
          -- ^ Defaults to @1@. Maximum number of rules to run in parallel, similar to @make --jobs=/N/@.
          --   For many build systems, a number equal to or slightly less than the number of physical processors
          --   works well. Use @0@ to match the detected number of processors (when @0@, 'getShakeOptions' will
  ```

  So I'm not sure it should count.

- Time

  I claim disqualified because it's not leading comma. It is actually aligning to the left margin, which could also be considered what I'm suggesting.

  ```hs
  -- | A TimeZone is a whole number of minutes offset from UTC, together with a name and a \"just for summer\" flag.
  data TimeZone = TimeZone {
      -- | The number of minutes offset from UTC. Positive means local time will be later 
  in the day than UTC.
      timeZoneMinutes :: Int,
      -- | Is this time zone just persisting for the summer?
      timeZoneSummerOnly :: Bool,
      -- | The name of the zone, typically a three- or four-letter acronym.
      timeZoneName :: String
  } deriving (Eq,Ord,Data, Typeable)
  ```

So you've got 4 examples: one in my favor, one in yours, one disqualified and one that's iffy :joy:

Reply:
This is the comment thread for records, not imports! Lens is doing what I want haha

Reply:
Are you proposing using a different style for exports vs records. I have a strong opinion that we can and should use a consistent guide, which is align to the `,`.

Reply:
No, I'm proposing aligning to the identifier on both, not the comma, since we aren't describing commas we are describing identifiers.

I'm fine with the style in this PR if I'm outvoted, I just think it looks bad and I was offering an alternative. I was also trying to dispel the idea that I was just "sprouting" this style out of nowhere.